### ISSUE TITLE
Report ValueTask returned by lambda passed to Task.Run

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#4290]
### Analyzer

**Diagnostic ID**: [CA2012](https://docs.microsoft.com/visualstudio/code-quality/ca2012): `Use ValueTasks correctly`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers?](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: 5.0.0-rc2.20453.1

### Describe the bug

CA2012 should be reported for `ValueTask<T>` returned from a delegate passed to `Task.Run`:

```
public async ValueTask<int> MethodAsync();

Task.Run(() => MethodAsync()); // <-- Should warn
```

### Steps To Reproduce

See description

**Comment by stephentoub** [comment]
That's a good find; we should improve this.  Ideally we'd add Task.Run overloads that take `Func<ValueTask>` as well in addition to the existing `Func<Task>`-based overloads, but it'd be a pretty big breaking change.

**Comment by stephentoub** [comment]
Technically, it'd be fine if someone did:
```C#
await await Task.Run(() => MethodAsync());
```
or
```C#
Task<ValueTask> t = Task.Run(() => MethodAsync());
ValueTask vt = await t;
await vt;
```
etc.  I'd say that's such an outlier as to not be worth catching, except such a situation is exactly why the `Task.Run(Func<Task>)` overload exists, to do that unwrapping for you.  And `await await Task.Run(() => MethodAsync());` is likely more efficient than `await Task.Run(async () => MethodAsync());`, so if such a case were hit, it's not ideal to recommend someone switch from the former to the latter.

So, this could be complicated.  It'd basically need to detect when:
- An anonymous function / lambda is passed to Task.Run<TResult>(Func<Task<TResult>>)
- That lambda directly returns a `ValueTask` or `ValueTask<T>`
- The `Task.Run` is either not awaited at all, or is awaited but the TResult is ignored (it could be made more complicated than that, but diminishing returns)

**Comment by sharwell** [comment]
I found this case because we had:

```
List<Task> tasks;

tasks.Add(Task.Run(() => MethodAsync());
```

And we changed `MethodAsync` from returning `Task<int>` to `ValueTask<int>`.

**Comment by stephentoub** [comment]
> The code still compiled but the meaning changed completely.

Yup.  My suggested heuristic wouldn't work on your case, either, but it's also possible in your case someone could have done:
```C#
foreach (ValueTask vt in tasks) await vt;
```
which is technically fine.  Hence the conundrum.  It's possible we say "this is in the weeds and warning is ok", because it'll be right more often than it's not... though I did try to make false positives here super, super rare, erring in general to have false negatives instead.

**Comment by stephentoub** [comment]
Thinking about this more...

I think I'm ok with flagging `Task.Run(() => returns ValueTask)`, regardless of what happens with the returned Task... but it should be its own diagnostic ID (and probably makes sense in its own analyzer, but that's an implementation detail).  This is really about the lack of a `Task.Run(Func<ValueTask>)` overload more than it is about consuming ValueTasks in a particular way.

**Comment by stephentoub** [comment]
And in that light, really this would just be an analyzer that warns when it sees an invocation of `Task.Run<ValueTask>`, regardless of whether it's a lambda or anonymous function or method group or delegate instance or whatever passed to it.

**Comment by amoerie** [comment]
We just hit a nasty bug that was caused by this. (Refactoring an existing method to return ValueTask instead of Task)
A compiler / analyzer warning would have been much appreciated. 😅
